go/chat/attachment_crypt.go

```diff
@@ -10,36 +16,83 @@ type Encrypter interface {
 	// Encrypt takes a plaintext reader and returns a ciphertext reader.
 	Encrypt(plaintext io.Reader) (ciphertext io.Reader)
 
-	// Key returns the ephemeral key that was used during Encrypt.
-	Key() []byte
+	// EncryptKey returns the ephemeral key that was used during Encrypt.
+	EncryptKey() []byte
+
+	// VerifyKey returns the public portion of the signing key that
+	// can be used for signature verification.
+	VerifyKey() []byte
 }
 
 type Decrypter interface {
-	// Decrypt takes a ciphertext reader and a key and returns a
-	// plaintext reader.
-	Decrypt(ciphertext io.Reader, key []byte) (plaintext io.Reader, err error)
+	// Decrypt takes a ciphertext reader, encryption and verify keys.
+	// It returns a plaintext reader.
+	Decrypt(ciphertext io.Reader, encKey, verifyKey []byte) (plaintext io.Reader)
 }
 
-// PassThrough implements Encrypter and Decrypter.  It does no encryption
-// or decryption.
-type PassThrough struct{}
+var nonce = &[signencrypt.NonceSize]byte{
+	0x10, 0x10, 0x10, 0x10, 0x20, 0x20, 0x20, 0x20,
+	0x30, 0x30, 0x30, 0x30, 0x40, 0x40, 0x40, 0x40,
+}
 
-func NewPassThrough() *PassThrough {
-	return &PassThrough{}
+type SignEncrypter struct {
+	encKey    signencrypt.SecretboxKey
+	signKey   signencrypt.SignKey
+	verifyKey signencrypt.VerifyKey
 }
 
-func (p *PassThrough) EncryptedLen(size int) int {
-	return size
+func NewSignEncrypter() (*SignEncrypter, error) {
+	enc, err := libkb.GenerateNaclDHKeyPair()
+	if err != nil {
+		return nil, err
+	}
+	sign, err := libkb.GenerateNaclSigningKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	var encKey [signencrypt.SecretboxKeySize]byte
+	copy(encKey[:], (*enc.Private)[:])
+	var signKey [ed25519.PrivateKeySize]byte
+	copy(signKey[:], (*sign.Private)[:])
+	var verifyKey [ed25519.PublicKeySize]byte
+	copy(verifyKey[:], sign.Public[:])
+
+	res := SignEncrypter{
+		encKey:    &encKey,
+		signKey:   &signKey,
+		verifyKey: &verifyKey,
+	}
+
+	return &res, nil
 }
 
-func (p *PassThrough) Encrypt(r io.Reader) io.Reader {
-	return r
+func (s *SignEncrypter) EncryptedLen(size int) int {
+	return signencrypt.GetSealedSize(size)
 }
 
-func (p *PassThrough) Key() []byte {
-	return nil
+func (s *SignEncrypter) Encrypt(r io.Reader) io.Reader {
+	return signencrypt.NewEncodingReader(s.encKey, s.signKey, nonce, r)
```

Comment:
This is the line of code that weirds me out the most in my [comment above](https://github.com/keybase/client/pull/4689/files#r84761498) about the way the interface is laid out.